include/revng/ADT/GenericGraph.h

```diff
@@ -10,8 +10,73 @@
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallVector.h"
 
+#include "revng/ADT/STLExtras.h"
 #include "revng/Support/Debug.h"
 
+/// GenericGraph is our implementation of a universal graph architecture.
+/// First and foremost, it's tailored for the revng codebase, but maybe you
+/// can find it useful as well, that's why it's released under the MIT License.
+///
+/// To use the Graph, you need to make you choice of the node type. We're
+/// currently supporting three architectures:
+///
+///   - ForwardNode - the most trivial one. It uses the least memory, but each
+///                   node only stores links with its successors. That makes
+///                   the inverse graph iteration expensive.
+///
+///   - BidirectionalNode - A simple double-linked node architecture.
+///                         Each node remembers links with both its successors
+///                         and predecessors. Take note, that for the sake of
+///                         implementation simplicity this node stores COPIES
+///                         of the labels. It's only suitable to be used with
+///                         cheap to copy labels which are never actually
+///                         mutated. There are plans on making them explicitly
+///                         immutable (TODO). So you can consider node mutation
+///                         a deprecated behaviour.
+///
+///    - MutableEdgeNode - The most complex node type we're supporting.
+///                        It is similar to BidirectionalNode except it stores
+///                        the label on the heap. It's slower and more memory
+///                        consuming but it allows for safe edge label
+///                        modification as well as controls that nodes and
+///                        edges are safely removed so no leftover parts are
+///                        ever present.
+///
+///    - The node you're going to write - No-one knows the needs of your
+///                                       betten than you. That's why the best
+///                                       datastructure for you is the one you
+///                                       are going to write. So just inherit
+///                                       one of our nodes, or even copy it and
+///                                       modify it so that it suits your graph
+///                                       as well as possible.
```

Comment:
I understand why you indent like this, but I think the effect here is excessive.

Reply:
Done.